File: ios/AusweisApp2SDKWrapper/Sources/AusweisApp2SDKWrapper/card/core/AusweisApp2/protocol/Messages.swift
/**
 * Copyright (c) 2020-2025 Governikus GmbH & Co. KG, Germany
 */

import Foundation

enum AA2Messages {
	static let msgAccessRights = "ACCESS_RIGHTS"
	static let msgAuth = "AUTH"
	static let msgCertificate = "CERTIFICATE"
	static let msgChangePin = "CHANGE_PIN"
	static let msgEnterPin = "ENTER_PIN"
	static let msgEnterNewPin = "ENTER_NEW_PIN"
	static let msgEnterPuk = "ENTER_PUK"
	static let msgEnterCan = "ENTER_CAN"
	static let msgInsertCard = "INSERT_CARD"
	static let msgBadState = "BAD_STATE"
	static let msgReader = "READER"
	static let msgInvalid = "INVALID"
	static let msgUnknowCommand = "UNKNOWN_COMMAND"
	static let msgInternalError = "INTERNAL_ERROR"
	static let msgStatus = "STATUS"
	static let msgInfo = "INFO"
	static let msgReaderList = "READER_LIST"
	static let msgPause = "PAUSE"
}

struct AA2Message: Decodable {
	let msg: String
	let error: String?
	let card: AA2Card?
	let result: AA2Result?
	let chat: AA2Chat?
	let aux: AA2Aux?
	let transactionInfo: String?
	let validity: AA2Validity?
	let description: AA2Description?
	let url: String?
	let success: Bool?
	let reason: String?
	let reader: AA2Reader?
	let readers: [AA2Reader]?
	let name: String?
	let insertable: Bool?
	let attached: Bool?
	let keypad: Bool?
	let workflow: String?
	let progress: Int?
	let state: String?
	let versionInfo: AA2VersionInfo?
	let cause: String?

	enum CodingKeys: String, CodingKey {
		case versionInfo = "VersionInfo"

		case msg, error, card, result, chat, aux, transactionInfo, validity
		case description, url, success, reader, workflow, progress, state
		case name, insertable, attached, keypad, readers, reason, cause
	}
}

struct AA2Description: Decodable {
	let issuerName: String
	let issuerUrl: String
	let purpose: String
	let subjectName: String
	let subjectUrl: String
	let termsOfUsage: String
}

struct AA2Validity: Decodable {
	let effectiveDate: String
	let expirationDate: String
}

struct AA2Chat: Decodable {
	let effective: [String]
	let optional: [String]
	let required: [String]
}

struct AA2Aux: Decodable {
	let ageVerificationDate: String?
	let requiredAge: String?
	let validityDate: String?
	let communityId: String?
}

struct AA2Card: Decodable {
	let deactivated: Bool?
	let inoperative: Bool?
	let retryCounter: Int?
}

struct AA2Reader: Decodable {
	let name: String
	let insertable: Bool
	let attached: Bool
	let keypad: Bool
	let card: AA2Card?
}

struct AA2Result: Decodable {
	let major: String?
	let minor: String?
	let url: String?
	let language: String?
	let description: String?
	let message: String?
	let reason: String?
}

struct AA2VersionInfo: Decodable {
	let name: String
	let implementationTitle: String
	let implementationVendor: String
	let implementationVersion: String
	let specificationTitle: String
	let specificationVendor: String
	let specificationVersion: String

	enum CodingKeys: String, CodingKey {
		case name = "Name"
		case implementationTitle = "Implementation-Title"
		case implementationVendor = "Implementation-Vendor"
		case implementationVersion = "Implementation-Version"
		case specificationTitle = "Specification-Title"
		case specificationVendor = "Specification-Vendor"
		case specificationVersion = "Specification-Version"
	}
}

File: android/sdkwrapper/src/main/java/com/governikus/ausweisapp/sdkwrapper/card/core/ausweisapp/protocol/Messages.kt
/**
 * Copyright (c) 2020-2026 Governikus GmbH & Co. KG, Germany
 */

package com.governikus.ausweisapp.sdkwrapper.card.core.ausweisapp.protocol

import com.squareup.moshi.Json

internal object Messages {
    const val MSG_ACCESS_RIGHTS = "ACCESS_RIGHTS"
    const val MSG_AUTH = "AUTH"
    const val MSG_BAD_STATE = "BAD_STATE"
    const val MSG_CERTIFICATE = "CERTIFICATE"
    const val MSG_CHANGE_PIN = "CHANGE_PIN"
    const val MSG_ENTER_CAN = "ENTER_CAN"
    const val MSG_ENTER_NEW_PIN = "ENTER_NEW_PIN"
    const val MSG_INFO = "INFO"
    const val MSG_ENTER_PIN = "ENTER_PIN"
    const val MSG_ENTER_PUK = "ENTER_PUK"
    const val MSG_INSERT_CARD = "INSERT_CARD"
    const val MSG_INTERNAL_ERROR = "INTERNAL_ERROR"
    const val MSG_INVALID = "INVALID"
    const val MSG_PAUSE = "PAUSE"
    const val MSG_READER = "READER"
    const val MSG_READER_LIST = "READER_LIST"
    const val MSG_STATUS = "STATUS"
    const val MSG_UNKNOWN_COMMAND = "UNKNOWN_COMMAND"
}

internal data class Message(
    val attached: Boolean?,
    val aux: Aux?,
    val card: Card?,
    val cause: String?,
    val chat: Chat?,
    val description: Description?,
    val error: String?,
    val insertable: Boolean?,
    val keypad: Boolean?,
    val msg: String?,
    val name: String?,
    val progress: Int?,
    val reader: Reader?,
    val readers: List<Reader>?,
    val reason: String?,
    val result: Result?,
    val state: String?,
    val success: Boolean?,
    val transactionInfo: String?,
    val url: String?,
    val validity: Validity?,
    @Json(name = "VersionInfo")
    val versionInfo: VersionInfo?,
    val workflow: String?,
)

internal data class Description(
    val issuerName: String,
    val issuerUrl: String,
    val purpose: String,
    val subjectName: String,
    val subjectUrl: String,
    val termsOfUsage: String,
)

internal data class Validity(
    val effectiveDate: String,
    val expirationDate: String,
)

internal data class Chat(
    val effective: List<String>,
    val optional: List<String>,
    val required: List<String>,
)

internal data class Aux(
    val ageVerificationDate: String?,
    val requiredAge: String?,
    val validityDate: String?,
    val communityId: String?,
)

internal data class Card(
    val inoperative: Boolean?,
    val deactivated: Boolean?,
    val retryCounter: Int?,
)

internal data class Reader(
    val name: String,
    val insertable: Boolean,
    val attached: Boolean,
    val keypad: Boolean,
    val card: Card?,
)

internal data class Result(
    val major: String?,
    val minor: String?,
    val language: String?,
    val description: String?,
    val message: String?,
    val reason: String?,
)

internal data class VersionInfo(
    @Json(name = "Name")
    val name: String,
    @Json(name = "Implementation-Title")
    val implementationTitle: String,
    @Json(name = "Implementation-Vendor")
    val implementationVendor: String,
    @Json(name = "Implementation-Version")
    val implementationVersion: String,
    @Json(name = "Specification-Title")
    val specificationTitle: String,
    @Json(name = "Specification-Vendor")
    val specificationVendor: String,
    @Json(name = "Specification-Version")
    val specificationVersion: String,
)
